### ISSUE TITLE
tizonia aac decoder: does not work on attached files

### ISSUE THREAD

**Issue by CapOM** [tizonia/tizonia-openmax-il#400]
Hello,

"OMX.Aratelia.audio_decoder.aac" does not work on the 2 files that I put all together in the attached archive.
It is both mp4 containing aac. First one is 7 channels, second is 2 channels (7 channles mixed/converted into 2 channels).

They play well with a browser or with gstreamer using the faad element wrapper:
gst-launch-1.0 filesrc location=7.1auditionOutLeader_v2_rtb.mp4 ! qtdemux ! aacparse ! faad ! audioconvert ! pulsesink

[aac.tar.gz](https://github.com/tizonia/tizonia-openmax-il/files/1522103/aac.tar.gz)

Replacing faad with omxaacdec leads to errors. The attached patch helps because NeAACDecInit(v1) return 0 instead of 2, whereas NeAACDecInit2 does not return the number of bytes. With that patch it runs only on the 2ch file but not audible (just noise but it is not that far from being correct).

That would be great if you could make aacdecprc.c handle these files.

Out of curiosity could you share a few files on which you tested it ? I will give it a go with omxaacdec.

Thx a lot
Julien

**Comment by juanrubio** [comment]
Hi Julien,

I'll have a look at those files.

Yes, I remember the tizonia OMX AAC decoder does not support all the AAC features very well. From my testing, I remember I could not make it to decode streams with some of the newer AAC features.

Is the gst faad component also based on libfaad? If that is so, then I'll definitely have a look at that code, because that would mean that my understading of the libfaad API is flawed.

This is the location of the streams that are being used in Travis. I have not done extensive testing of the AAC decoder (like using of the conformance testing stream samples out there, etc)
http://www.aratelia.com/tizonia/tizonia-test-media.tgz

Finally, is there a patch you were meaning to attach?

**Comment by CapOM** [comment]
Ah sorry, here is the diff :

--- a/plugins/aac_decoder/src/aacdecprc.c
+++ b/plugins/aac_decoder/src/aacdecprc.c
@@ -334,7 +334,7 @@ static void store_stream_metadata (aacdec_prc_t *ap_prc)
 static OMX_ERRORTYPE init_aac_decoder (aacdec_prc_t *ap_prc)
 {
   OMX_ERRORTYPE rc = OMX_ErrorInsufficientResources;
-  long nbytes = 0;
+  int8_t init_result = 0;
   OMX_BUFFERHEADERTYPE *p_in = tiz_filter_prc_get_header (
       ap_prc, ARATELIA_AAC_DECODER_INPUT_PORT_INDEX);
 
@@ -363,16 +363,16 @@ static OMX_ERRORTYPE init_aac_decoder (aacdec_prc_t *ap_prc)
      (skip_id3_tag may have modified it */
 
   /* Initialise the library using one of the initialization functions */
-  nbytes = NeAACDecInit (ap_prc->p_aac_dec_,
+  init_result = NeAACDecInit2 (ap_prc->p_aac_dec_,
                          tiz_buffer_get (ap_prc->p_store_),
                          tiz_buffer_available (ap_prc->p_store_),
                          &(ap_prc->samplerate_), &(ap_prc->channels_));
 
-  if (nbytes < 0)
+  if (init_result < 0)
     {
       TIZ_ERROR (handleOf (ap_prc),
                  "[%s] : libfaad decoder initialisation failure (%s)",
-                 tiz_err_to_str (rc), NeAACDecGetErrorMessage (nbytes));
+                 tiz_err_to_str (rc), NeAACDecGetErrorMessage (tiz_buffer_available (ap_prc->p_store_)));
     }
   else
     {
@@ -380,7 +380,7 @@ static OMX_ERRORTYPE init_aac_decoder (aacdec_prc_t *ap_prc)
       tiz_check_omx (update_pcm_mode (ap_prc, ap_prc->samplerate_,
                                           ap_prc->channels_));
       /* We will skip this many bytes the next time we read from this buffer */
-      tiz_buffer_advance (ap_prc->p_store_, nbytes);
+      tiz_buffer_advance (ap_prc->p_store_, tiz_buffer_available (ap_prc->p_store_));
       TIZ_DEBUG (handleOf (ap_prc), "samplerate [%d] channels [%d]",
                  ap_prc->samplerate_, (int)ap_prc->channels_);
       store_stream_metadata (ap_prc);
--

**Comment by CapOM** [comment]
[0001-wip.patch.tar.gz](https://github.com/tizonia/tizonia-openmax-il/files/1522224/0001-wip.patch.tar.gz)

**Comment by CapOM** [comment]
Have a look to : https://cgit.freedesktop.org/gstreamer/gst-plugins-bad/tree/ext/faad/gstfaad.c

**Comment by CapOM** [comment]
gstomxaacdec works on both tizonia-test-media/aac/riff_rock.aac and tizonia-test-media/aac/strum12str.aac without the attached patch I attached, but with this change:

In plugins/aac_decoder/src/aacdec.c, put pcmmode.eEndian = OMX_EndianLittle; instead of OMX_EndianBig . 

With both patches it can play aac_2ch.mp4 so the big endian /little endian was the explanation for the noise I mentioned above.

So there are at least 2 issues, the endianness and needs to call NeAACDecInit2 or NeAACDecInit depending on some heuristic or try NeAACDecInit2 first and if fails try NeAACDecInit, something like that.

But for 7.1auditionOutLeader_v2_rtb.mp4 it looks it needs more than that.

Thx!

**Comment by juanrubio** [comment]
Hi Julien (@CapOM)

I remember now why I did not use at the time the `NeAACDecInit2` call in the Tizonia AAC decoder. The reason is that this api is used when the first buffer in the pipeline contains the AAC AudioSpecificConfig element. 

```
/* Init the library using a DecoderSpecificInfo */
char NEAACDECAPI NeAACDecInit2(NeAACDecHandle hDecoder,
```

The behaviour when there is a decoder specific info element in OpenMAX IL was specified in both 1.1.2  in 1.2):

```

/** Codec Config Buffer Flag:                                                                                                                                               
* OMX_BUFFERFLAG_CODECCONFIG is an optional flag that is set by an                                                                                                          
* output port when all bytes in the buffer form part or all of a set of                                                                                                     
* codec specific configuration data.  Examples include SPS/PPS nal units                                                                                                    
* for OMX_VIDEO_CodingAVC or AudioSpecificConfig data for                                                                                                                   
* OMX_AUDIO_CodingAAC.  Any component that for a given stream sets                                                                                                          
* OMX_BUFFERFLAG_CODECCONFIG shall not mix codec configuration bytes                                                                                                        
* with frame data in the same buffer, and shall send all buffers                                                                                                            
* containing codec configuration bytes before any buffers containing                                                                                                        
* frame data that those configurations bytes describe.                                                                                                                      
* If the stream format for a particular codec has a frame specific                                                                                                          
* header at the start of each frame, for example OMX_AUDIO_CodingMP3 or                                                                                                     
* OMX_AUDIO_CodingAAC in ADTS mode, then these shall be presented as                                                                                                        
* normal without setting OMX_BUFFERFLAG_CODECCONFIG.                                                                                                                        
 * @ingroup buf                                                                                                                                                             
 */
#define OMX_BUFFERFLAG_CODECCONFIG 0x00000080
```

Therefore, since Tizonia does not yet have an mp4 demuxer (see #392), the only streams I could support at the time were those that can be decoded with
```
/* Init the library based on info from the AAC file (ADTS/ADIF) */
long NEAACDECAPI NeAACDecInit(NeAACDecHandle hDecoder,
```

So before tizonia's AAC decoder can support both modes of operation, and decode aac_2ch.mp4, ideally I should try to implement the mp4 demuxer to get a testing pipeline to play with.

Do you know if the gst-omx outputs this OMX_BUFFERFLAG_CODECCONFIG in its first buffer?

**Comment by juanrubio** [comment]
It is possible to do the AudioSpecificConfig work on the AAC decoder without the mp4 demuxer. The binary file reader with small modifications can be used to get the job done. I cannot promise to get this delivered right away, since I would like to resolve #393 as soon as possible.

**Comment by CapOM** [comment]
>> "Do you know if the gst-omx demuxer outputs OMX_BUFFERFLAG_CODECCONFIG in its first buffer?"

Yes it does https://cgit.freedesktop.org/gstreamer/gst-omx/tree/omx/gstomxaudiodec.c#n1175
